aspnet/fundamentals/middleware.rst

```diff
@@ -1,8 +1,176 @@
+
+.. note:: **Remember:** the order in which you arrange your ``Use[Middleware]`` statements in your application's ``Configure`` method is very important. Be sure you have a good understanding of how your application's request pipeline will behave in various scenarios.
+
+The simplest possible ASP.NET application sets up a single request delegate that handles all requests. In this case, there isn't really a request "pipeline", so much as a single anonymous function that is called in response to every HTTP request.
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 23-26
+	:dedent: 12
+
+It's important to realize that request delegate, as written here, will terminate the pipeline, regardless of other calls to ``App.Run`` that you may include. In the following example, only the first delegate ("Hello, World!") will be executed and displayed.
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 20-31
+	:emphasize-lines: 5
+	:dedent: 8
+
+You chain multiple request delegates together making a different call, with a ``next`` parameter representing the next delegate in the pipeline. Note that just because you're calling it "next" doesn't mean you can't perform actions both before and after the next delegate, as this example demonstrates:
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 34-49
+	:emphasize-lines: 5,8,14
+	:dedent: 8
+
+.. note:: This ``ConfigureLogInline`` method is called when the application is run with an environment set to ``LogInline``. Learn more about :doc:`environments`. We will be using variations of ``Configure[Environment]`` to show different options in the rest of this article.
+
+In the above example, the call to ``await next.Invoke()`` will call into the delegate on line 14. The client will receive the expected response ("Hello from LogInline"), and the server's console output includes both the before and after messages, as you can see here:
+
+.. image:: middleware/_static/console-loginline.png
+
+Run, Map, and Use
+^^^^^^^^^^^^^^^^^
+
+You configure the HTTP pipeline using the `extensions <https://github.com/aspnet/HttpAbstractions/tree/dev/src/Microsoft.AspNet.Http.Abstractions/Extensions>`_ ``Run``, ``Map``, and ``Use``. By convention, the ``Run`` method is simply a shorthand way of adding middleware to the pipeline that doesn't call any other middleware (that is, it will not call a ``next`` request delegate). The following two examples (one using ``Run`` and the other ``Use``) are equivalent to one another, since the second one doesn't use its ``next`` parameter:
```

Comment:
Note that Run is a convention, not just a single extension. Run[Middleware] extensions may be provided by some components that should only run at the end of the pipeline.